providers/caching.go

```diff
@@ -8,58 +8,49 @@ import (
 	"github.com/libp2p/go-libp2p-kad-dht/providers"
 	"github.com/libp2p/hydra-booster/metrics"
 	"go.opencensus.io/stats"
-	"go.opencensus.io/tag"
 )
 
-// CachingProviderStore wraps a providerstore, and finds providers for CIDs that were requested
-// but not found in the underlying providerstore, and then caches them in the providerstore.
+// CachingProviderStore checks the ReadProviderStore for providers. If no providers are returned,
+// then the Finder is used to find providers, which are then added to the WriteProviderStore.
 type CachingProviderStore struct {
-	Delegate        providers.ProviderStore
-	Finder          ProvidersFinder
-	Router          ReadContentRouting
-	ProvidersToFind int
-	log             logging.EventLogger
+	ReadProviderStore  providers.ProviderStore
+	WriteProviderStore providers.ProviderStore
+	Finder             ProvidersFinder
+	Router             ReadContentRouting
+	log                logging.EventLogger
 }
 
-func NewCachingProviderStore(delegate providers.ProviderStore, finder ProvidersFinder, router ReadContentRouting) *CachingProviderStore {
+func NewCachingProviderStore(getDelegate providers.ProviderStore, addDelegate providers.ProviderStore, finder ProvidersFinder, router ReadContentRouting) *CachingProviderStore {
 	return &CachingProviderStore{
-		Delegate: delegate,
-		Finder:   finder,
-		Router:   router,
-		log:      logging.Logger("hydra/providersn"),
+		ReadProviderStore:  getDelegate,
+		WriteProviderStore: addDelegate,
+		Finder:             finder,
+		Router:             router,
+		log:                logging.Logger("hydra/providers"),
 	}
 }
 
 func (s *CachingProviderStore) AddProvider(ctx context.Context, key []byte, prov peer.AddrInfo) error {
-	return s.Delegate.AddProvider(ctx, key, prov)
+	return s.WriteProviderStore.AddProvider(ctx, key, prov)
 }
 
 // GetProviders gets providers for the given key from the providerstore.
 // If the providerstore does not have providers for the key, then the ProvidersFinder is queried and the results are cached.
 func (d *CachingProviderStore) GetProviders(ctx context.Context, key []byte) ([]peer.AddrInfo, error) {
-	addrInfos, err := d.Delegate.GetProviders(ctx, key)
+	addrInfos, err := d.ReadProviderStore.GetProviders(ctx, key)
 	if err != nil {
 		return addrInfos, err
 	}
 
 	if len(addrInfos) > 0 {
-		recordPrefetches(ctx, "local")
 		return addrInfos, nil
 	}
 
 	return nil, d.Finder.Find(ctx, d.Router, key, func(ai peer.AddrInfo) {
```

Comment:
This is probably not worth the effort, but for future reference. It is bad style to reuse the error return value for two different purposes. Above, returned errors indicate errors in the providerstores. Here, errors indicate errors in the finder. Considering that the finder is an async functionality, independent of the GetProviders execution path, its error should be logged, not returned.

Reply:
Otherwise this function seems to be a correct implementation of the logic we discussed in colo.

Reply:
I haven't heard of that style guideline before, if a method is supposed to do two things, and it can't do one of them, then it's normal to return an error, regardless of which one failed. The finder is best effort, so an error from the finder here doesn't represent an error in the finding itself, it means that we couldn't even _try_ to find (e.g. in async case, that the work couldn't be queued for some reason). My intention was for the CachingProviderStore to not know nor care that things are happening async, but given that we don't have a sync version of this, I can see how that just adds confusion, so I can rename things a bit here to clarify.